### PR TITLE
sql: backward-compatibility for SHOW RANGES / crdb_internal.ranges

### LINKED ISSUE #97861: sql: add a NOTICE for uses of SHOW RANGES / crdb_internal.ranges
Following #[93644](https://github.com/cockroachdb/cockroach/pull/93644) we may want to inform users who are attempting to use SHOW RANGES / crdb_internal.ranges using a NOTICE message about the changes.

Jira issue: CRDB-24927
Epic: CRDB-24928

### LINKED ISSUE #99073: sql: explore a configurable switch to flip between old/new behavior in `crdb_internal.ranges{,no_leases}` and `SHOW RANGES`
Requested by @nvanbenschoten and @mwang1026 
Epic: CRDB-24928

We are interested to explore options to make the v22.x behavior for these range inspection facilities available _conditionally_ in a v23.1 deployment, up to the point the operator is ready to switch; in a way that would be deprecated in v23.1 (and removed in v23.2.)

Jira issue: CRDB-25689

### PR DESCRIPTION
**Pull request by knz**
Requested by @mwang1026  and @nvanbenschoten .

Fixes #97861.
Fixes #99073.
Needed for #98820.
Epic: CRDB-24928

See the individual commits for details; exceprt from the 2nd commit:

TLDR: the pre-v23.1 behavior of SHOW RANGES and
`crdb_internal.ranges{_no_leases}` is made available conditional on a
new cluster setting `sql.show_ranges_deprecated_behavior.enabled`.

It is set to true by default in v23.1, however any use of the feature
will prompt a SQL notice with the following message:

```
NOTICE: attention! the pre-23.1 behavior of SHOW RANGES and crdb_internal.ranges{,_no_leases} is deprecated!
HINT: Consider enabling the new functionality by setting 'sql.show_ranges_deprecated_behavior.enabled' to 'false'.
The new SHOW RANGES statement has more options. Refer to the online
documentation or execute 'SHOW RANGES ??' for details.
```

The deprecated behavior is also disabled in the test suite and by default in
`cockroach demo`, i.e. the tests continue to exercise the new
behavior.

Release note (backward-incompatible change): The pre-v23.1 output
produced by `SHOW RANGES`, `crdb_internal.ranges`,
`crdb_internal.ranges_no_leases` is deprecated. The new interface and
functionality for `SHWO RANGES` can be enabled by changing the cluster
setting `sql.show_ranges_deprecated_behavior.enabled` to `false`. It
will become default in v23.2.

### CODE REVIEW

**Review by cucaroach — APPROVED** [review]
:lgtm:

**Review by fqazi — APPROVED** [review]
Nothing major just minor nits

> Reviewed 4 of 4 files at r1, 25 of 25 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/98979#-:-NRAk0xxECFuZpgI4ssF:b-vtok3q)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @irfansharif and @knz)

___
*[`pkg/sql/crdb_internal_ranges_deprecated.go` line 34 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/98979#-NRAhWcj3s-0-hSMy7cO:-NRAhWck4xkdKaQnbc2U:bnx8a35) ([raw file](https://github.com/cockroachdb/cockroach/blob/c575203e31331b8809dedf2bb82b89b60fb0b0f6/pkg/sql/crdb_internal_ranges_deprecated.go#L34)):*
> ```Go
> // version of `crdb_internal.ranges`. See the package-level doc
> // of package `deprecatedshowranges` for details.
> var crdbInternalRangesViewDEPRECATED = virtualSchemaView{
> ```

Maybe just add a version number instead and have a TODO for removal
___
*[`pkg/sql/virtual_schema.go` line 445 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/98979#-NRAhxw40-CuqxM4P8Cy:-NRAhxw40-CuqxM4P8Cz:be1pz5t) ([raw file](https://github.com/cockroachdb/cockroach/blob/c575203e31331b8809dedf2bb82b89b60fb0b0f6/pkg/sql/virtual_schema.go#L445)):*
> ```Go
> 	case id == catconstants.CrdbInternalRangesViewID && deprecatedshowranges.EnableDeprecatedBehavior(context.TODO(), vs.st, nil):
> 		entry = vs.schemasByID[catconstants.CrdbInternalID].extraRangesDef
> 	case id == catconstants.CrdbInternalRangesNoLeasesTableID && deprecatedshowranges.EnableDeprecatedBehavior(context.TODO(), vs.st, nil):
> ```

Can we pipe in a context too?
___
*[`pkg/sql/virtual_schema.go` line 1000 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/98979#-NRAji8BATETc3pqwP4b:-NRAji8BATETc3pqwP4c:bqllpdf) ([raw file](https://github.com/cockroachdb/cockroach/blob/c575203e31331b8809dedf2bb82b89b60fb0b0f6/pkg/sql/virtual_schema.go#L1000)):*
> ```Go
> 			// Remove in v23.2.
> 			st:                     st,
> 			extraRangesDef:         extra1,
> ```

Instead of extra just label them as deprecated or with the version

**Review by knz — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/98979)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @fqazi and @irfansharif)

___
*[`pkg/sql/crdb_internal_ranges_deprecated.go` line 34 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/98979#-NRAhWcj3s-0-hSMy7cO:-NRSm4Bz3zyn9-4DUpNG:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c575203e31331b8809dedf2bb82b89b60fb0b0f6/pkg/sql/crdb_internal_ranges_deprecated.go#L34)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

Maybe just add a version number instead and have a TODO for removal
</blockquote></details>

Done.
___
*[`pkg/sql/virtual_schema.go` line 445 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/98979#-NRAhxw40-CuqxM4P8Cy:-NRSln15-P2PKYVAkQv1:b-lpazjg) ([raw file](https://github.com/cockroachdb/cockroach/blob/c575203e31331b8809dedf2bb82b89b60fb0b0f6/pkg/sql/virtual_schema.go#L445)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

Can we pipe in a context too?
</blockquote></details>

I tried - but then I need to change the interface in `catalog` which then requires me to change `(*byIDLookupContext).lookupVirtual` in `catalog/desc`, which then requires me to change all the other `lookupFunc`s. That's too much change for this PR.
___
*[`pkg/sql/virtual_schema.go` line 1000 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/98979#-NRAji8BATETc3pqwP4b:-NRSm26Y7QHIXyRJTuFd:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c575203e31331b8809dedf2bb82b89b60fb0b0f6/pkg/sql/virtual_schema.go#L1000)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

Instead of extra just label them as deprecated or with the version
</blockquote></details>

Done.

**Comment by knz** [comment]
TYFRs!

bors r=cucaroach,fqazi

**Comment by mikeCRL** [comment]
@knz Is there a reason to keep this in Backward-Incompatible Changes rather than Deprecations? It looks like we maintain the behavior of v22.2 by default in v23.1, though it is deprecated in v23.1 and will change by default in v23.2.

**Comment by knz** [comment]
> Is there a reason to keep this in Backward-Incompatible Changes rather than Deprecations? 

I chose that category to increase the chance the doc writer who sees it will go back to the _earlier_ release note in the release cycle that was marking the output as deprecated, and revise it.

Customers are going to look at the _combined_ list of deprecation notices across the entire 23.1 release and we need to build a consistent messaging, by editing the earlier release note.

**Comment by mikeCRL** [comment]
@knz I see. Thanks. Indeed I'd aggregated all prior Backward-Incompatible Changes and that captured all the `SHOW RANGES` changes. Other than bumping up the last `SHOW RANGES` bullet so that it falls directly after the others, would you recommend any other changes to what we have in the draft for that section? ([Preview](https://deploy-preview-16951--cockroachdb-docs.netlify.app/docs/releases/v23.1.html#v23-1-0-backward-incompatible-changes) / [PR](https://github.com/cockroachdb/docs/pull/16951/)

**Comment by knz** [comment]
i added some comments to that PR thanks
